File: visitor-pattern/python/src/ConcreteElementB.py
# -*- coding: utf-8 -*-
"""
@author: jarry
"""

from src.Element import Element


# 具体的元素实现者B
class ConcreteElementB(Element):

    def __init__(self):
        Element.__init__(self)
        self.name = 'Keyboard Element'

    def accept(self, visitor):
        print('ConcreteElementB::accept() [visitor.class = ' +
              visitor.__class__.__name__ + ' visitor.name = ' +
              visitor.get_name() + ']')
        visitor.visit(self)

    def operate(self):
        print('ConcreteElementB::operate() [' + self.get_name() + ']')

    def get_name(self):
        return self.name

File: visitor-pattern/java/src/ConcreteElementB.java
package src;

// 具体的元素实现者B
public class ConcreteElementB extends Element {
  private String name = "Keyboard Element";

  @Override
  public void accept(Visitor visitor) {
    System.out.println(
        "ConcreteElementB::accept() [visitor.class = " + visitor.getClass().getSimpleName() + " visitor.name = "
            + visitor.getName() + "]");
    visitor.visit(this);
  }

  public void operate() {
    System.out.println("ConcreteElementB::operate() [" + this.getName() + "]");
  }

  public String getName() {
    return this.name;
  }
}
